src/xdplwf/recv.c

```diff
@@ -1368,156 +1368,182 @@ XdpGenericReceiveEnqueueTxNbl(
-            //
-            if (FragmentCount + 1ui32 > RxQueue->FragmentLimit) {
-                //
-                // Generic XDP reserves a single contiguous buffer for
-                // linearization; if a NB contains more than the maximum number
-                // of fragments (MDLs), copy the contents of the entire MDL
-                // chain into one contiguous buffer.
-                //
-                // If the contiguous buffer is already in use, flush the queue
-                // first.
-                //
-                if (RxQueue->FragmentBufferInUse) {
-                    return;
-                }
+    FragmentExtension = XdpGetFragmentExtension(Frame, &RxQueue->FragmentExtension);
+    FragmentExtension->FragmentBufferCount = FragmentCount;
 
-                if (!XdpGenericReceiveLinearizeNb(RxQueue, *Nb)) {
-                    STAT_INC(&RxQueue->PcwStats, LinearizationFailures);
-                    goto Next;
-                }
+    //
+    // Store the original NB address so uninspected frames (e.g. those where
+    // virtual mappings failed) can be identified and dropped later.
+    //
+    InterfaceExtension =
+        XdpGetFrameInterfaceContextExtension(Frame, &RxQueue->FrameInterfaceContextExtension);
+    InterfaceExtension->Nb = Nb;
 
-                //
-                // The NB was converted to a single, virtually-contiguous buffer,
-                // so abandon fragmentation and proceed to XDP inspection.
-                //
-                FragmentCount = 0;
-                break;
-            }
+    //
+    // The NB has successfully been converted to XDP descriptors, so commit
+    // the descriptors to the XDP rings.
+    //
+    FragmentRing->ProducerIndex += FragmentCount;
+    FrameRing->ProducerIndex++;
 
-            //
-            // Reserve XDP descriptor space for this MDL. If space does not exist,
-            // flush existing descriptors to create space.
-            //
-            if (++FragmentCount > XdpRingFree(FragmentRing)) {
-                return;
-            }
+    *NbAddedToRing = TRUE;
+    return;
+}
 
-            Buffer =
-                XdpRingGetElement(
-                    FragmentRing,
-                    (FragmentRing->ProducerIndex + FragmentCount - 1) & FragmentRing->Mask);
-
-            Buffer->DataOffset = 0;
-            Buffer->DataLength = min(Mdl->ByteCount, DataLength);
-            Buffer->BufferLength = Mdl->ByteCount;
-            DataLength -= Buffer->DataLength;
-
-            SystemVa = XdpGetVirtualAddressExtension(Buffer, &RxQueue->BufferVaExtension);
-            SystemVa->VirtualAddress =
-                MmGetSystemAddressForMdlSafe(Mdl, LowPagePriority | MdlMappingNoExecute);
-            if (SystemVa->VirtualAddress == NULL || XdpLwfFaultInject()) {
-                STAT_INC(&RxQueue->PcwStats, MappingFailures);
-                goto Next;
-            }
-        }
+static
+BOOLEAN
```

Comment:
Same comment here about returning `BOOLEAN` vs using a named out parameter, though perhaps "success" of the preinspection routine implies the inspection routine should be invoked?

Reply:
That was how I was originally seing it. I don't mind an out parameter either if that feels more natural to you, otherwise I'll just leave it as it is.